pkg/daemon/certificate_writer.go

```diff
@@ -373,3 +425,164 @@ func writeToDisk(pathToData map[string][]byte) error {
 	}
 	return nil
 }
+
+// Reconciles and merges the secrets provided by the ControllerConfig along
+// with any mounted image pull secrets that the MCO may have configured the MCD
+// to use in a layered OS image scenario.
+func reconcileOSImageRegistryPullSecretData(node *corev1.Node, controllerCfg *mcfgv1.ControllerConfig, secretDirPath string) ([]byte, error) {
+	// First, get all of the node-roles that this node might have so we can
+	// resolve them to MachineConfigPools.
+	nodeRoles := getNodeRoles(node)
+
+	// This isn't likely to happen, but a guard clause here is nice.
+	if len(nodeRoles) == 0 {
+		return nil, fmt.Errorf("node %s has no node-role.kubernetes.io label", node.Name)
+	}
+
+	// Next, we need to read the secret from disk, if we can find one.
+	mountedSecret, err := readMountedSecretByNodeRole(nodeRoles, secretDirPath)
+	if err != nil {
+		return nil, err
+	}
+
+	// If there are no mounted secrets, just use the internal image pull secrets
+	// as-is.
+	if mountedSecret == nil {
+		klog.V(4).Infof("Did not find a mounted secret")
+		return controllerCfg.Spec.InternalRegistryPullSecret, nil
+	}
+
+	// Next, we need to merge the secret we just found with the contents of the
+	// InternalImagePullSecret field on the ControllerConfig object, ensuring
+	// that the values from the mounted secret take precedent over any values
+	// provided by the InternalImagePullSecret field.
+	merged, err := mergeMountedSecretsWithControllerConfig(mountedSecret, controllerCfg)
+	if err != nil {
+		return nil, fmt.Errorf("could not merge on-disk secrets with ControllerConfig: %w", err)
+	}
+
+	return merged, nil
+}
+
+// This iterates through all of the node roles until it finds a secret that
+// matches. We do this because we have to mount all of the secrets for all of
+// the MachineOSConfigs into each MCD pod. Additionally, we want to support the
+// case where someone uses wither a .dockercfg or .dockerconfigjson style
+// secret. However, it is possible that a different MachineOSConfig can specify
+// a different image pull secret. And it is also possible that a node can
+// belong to two MachineConfigPools.
+func readMountedSecretByNodeRole(nodeRoles []string, secretDirPath string) ([]byte, error) {
+	// If this directory does not exist, it means that the MCD DaemonSet does not
+	// have the secrets mounted at all. In this case, just return an empty byte
+	// array.
+	_, err := os.Stat(secretDirPath)
+	if errors.Is(err, os.ErrNotExist) {
+		klog.Infof("Path %s does not exist", secretDirPath)
```

Comment:
So if I am understanding this correctly, this shouldn't exist unless you have OS builds turned on and the secret is mounted to the daemonset?

According to the logs (e.g. https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openshift_machine-config-operator/4395/pull-ci-openshift-machine-config-operator-master-e2e-gcp-op/1801305805287329792/artifacts/e2e-gcp-op/gather-extra/artifacts/pods/openshift-machine-config-operator_machine-config-daemon-cc462_machine-config-daemon.log) this seems to be the case (which I think is normal). We can probably also turn down the verbosity here

Reply:
Agreed. I'll put it behind a `V(4)`.

Reply:
That is correct. If the root path here does not exist, we don't go any further with respect to the secret reconciliation.